js/modules/k6/http/async_request_test.go

```diff
@@ -0,0 +1,290 @@
+			require.NoError(t, err)
+			promiseRejected := ts.runtime.VU.Runtime().Get("promiseRejected")
+			exceptionThrown := ts.runtime.VU.Runtime().Get("exceptionThrown")
+			require.NotNil(t, promiseRejected)
+			require.True(t, promiseRejected.ToBoolean())
+			require.Nil(t, exceptionThrown)
+			assert.Contains(t, promiseRejected.ToString(), expErr)
+		})
+
+		t.Run("throw=false", func(t *testing.T) {
+			t.Parallel()
+			ts := newTestCase(t)
+			rt := ts.runtime.VU.Runtime()
+			state := ts.runtime.VU.State()
+			state.Options.Throw.Bool = false
+			defer func() { state.Options.Throw.Bool = true }()
+
+			hook := logtest.NewLocal(state.Logger)
+			defer hook.Reset()
+
+			js := `
+                var r = await http.asyncRequest("GET", "https:// test.k6.io");
+                globalThis.ret = {error: r.error, error_code: r.error_code};
+			`
+			err := runOnEventLoop(ts.runtime, js)
+			require.NoError(t, err)
+			ret := rt.GlobalObject().Get("ret")
+			var retobj map[string]interface{}
+			var ok bool
+			if retobj, ok = ret.Export().(map[string]interface{}); !ok {
+				require.Fail(t, "got wrong return object: %#+v", retobj)
+			}
+			require.Equal(t, int64(1020), retobj["error_code"])
+			require.Equal(t, expErr, retobj["error"])
+
+			logEntry := hook.LastEntry()
+			require.NotNil(t, logEntry)
+			assert.Equal(t, logrus.WarnLevel, logEntry.Level)
+			err, ok = logEntry.Data["error"].(error)
+			require.True(t, ok)
+			assert.ErrorContains(t, err, expErr)
+			assert.Equal(t, "Request Failed", logEntry.Message)
+		})
+
+		t.Run("throw=false,nopanic", func(t *testing.T) {
+			t.Parallel()
+			ts := newTestCase(t)
+			rt := ts.runtime.VU.Runtime()
+			state := ts.runtime.VU.State()
+			state.Options.Throw.Bool = false
+			defer func() { state.Options.Throw.Bool = true }()
+
+			hook := logtest.NewLocal(state.Logger)
+			defer hook.Reset()
+
+			js := `
+                var r = await http.asyncRequest("GET", "https:// test.k6.io");
+                r.html();
+                r.json();
+                globalThis.ret = r.error_code; // not reached because of json()
+			`
+			err := runOnEventLoop(ts.runtime, js)
+			ret := rt.GlobalObject().Get("ret")
+			require.Error(t, err)
+			assert.Nil(t, ret)
+			assert.Contains(t, err.Error(), "unexpected end of JSON input")
+
+			logEntry := hook.LastEntry()
+			require.NotNil(t, logEntry)
+			assert.Equal(t, logrus.WarnLevel, logEntry.Level)
+			err, ok := logEntry.Data["error"].(error)
+			require.True(t, ok)
+			assert.ErrorContains(t, err, expErr)
+			assert.Equal(t, "Request Failed", logEntry.Message)
+		})
+	})
+
+	t.Run("Unroutable", func(t *testing.T) {
+		t.Parallel()
+		ts := newTestCase(t)
+		err := runOnEventLoop(ts.runtime, `
+            try {
+                http.asyncRequest("GET", "http://sdafsgdhfjg/").catch((e) => globalThis.promiseRejected = e )
+            } catch (e) {
+                globalThis.exceptionThrown = e
+            }`)
+		expErr := "lookup sdafsgdhfjg"
```

Comment:
I got a failure from this test the first time I ran it:

```
=== CONT  TestAsyncRequestErrors/Unroutable
    async_request_test.go:288:
                Error Trace:    /home/ivan/Projects/grafana/k6/js/modules/k6/http/async_request_test.go:288
                Error:          "Get \"http://sdafsgdhfjg/\": request timeout" does not contain "lookup sdafsgdhfjg"
                Test:           TestAsyncRequestErrors/Unroutable
--- FAIL: TestAsyncRequestErrors (0.00s)
```

But then couldn't reproduce it after that. :confused:

Do you have an idea why it might've happened that one time?

Reply:
I see absolutely no way in which this is possible :thinking: :( 

More importantly I don't see anyway in which the original not async test will not be hitting it as well. If anything it likely can hit it easier as it is reusing a bunch of stuff from other tests while this one gets a clean slate each time. 

I can't even imagine how you can get "request timeout", have a 0s test run and only this test to fail. Even if you go change something or have not checked out cleanly I would expect a bunch of other tests to get basically the same error. 

The only explanation is that something within net/http does something wrong and mistakes the error which seems ... very unlikely, but I have nothing else. 

WDYT? Is there anything that you can think off that you did or could've happened

Reply:
Yeah, no idea either :confused:

I just ran `go test -v -race -failfast -count=1 -run 'TestAsyncRequest' ./js/modules/k6/http`, and then ran it again and it passed, and then with `-count=1000` and it kept passing until I stopped it.

I did clean my go mod cache recently, so maybe something was "warming up" that first time, but have no real explanation for it. Ah well, no need to dig into it if we can't reproduce it.